crates/ark/src/modules/positron/graphics.R

```diff
@@ -222,6 +193,142 @@ with_graphics_device <- function(
     expr
 }
 
+use_ragg <- local({
+    # Only check global option once per session
+    USE_RAGG <- NULL
+
+    function() {
+        if (is.null(USE_RAGG)) {
+            USE_RAGG <<- init_use_ragg()
+        }
+
+        USE_RAGG
+    }
+})
+
+init_use_ragg <- function() {
+    option <- getOption("ark.ragg", default = TRUE)
+
+    if (!isTRUE(option)) {
+        # Bail on any non-`TRUE` option value
+        return(FALSE)
+    }
+
+    if (!.ps.is_installed("ragg", minimum_version = "1.4.0")) {
+        # Need support for `agg_record()`
+        return(FALSE)
+    }
+
+    TRUE
+}
+
+#' Create a device to shadow
+#'
+#' For both ragg and png, we are hopeful that providing a `res` of the default
+#' resolution should not affect the render time results much (since we are just
+#' writing display list instructions), even if at render time we can actually
+#' support a `pixel_ratio` of 2x the default resolution. We simply don't know
+#' the pixel ratio at this point.
+device_shadow <- function() {
+    if (use_ragg()) {
+        # For the shadow ragg device, we use a special device that only captures
+        # the display list, it doesn't actually do any rendering!
+        ragg::agg_record(
+            res = default_resolution_in_pixels_per_inch()
+        )
+    } else {
+        # For the shadow png device, we need a dummy file to write to,
+        # even though we never look at it. We only utilize the png device for
+        # the action of recording the display list.
+        directory <- render_directory()
+        filename <- file.path(directory, "dummy-plot.png")
+
+        withCallingHandlers(
+            grDevices::png(
+                filename = filename,
+                type = default_device_type(),
+                res = default_resolution_in_pixels_per_inch()
+            ),
+            warning = function(w) {
+                stop("Error creating graphics device: ", conditionMessage(w))
+            }
+        )
+    }
+}
+
+device_png <- function(filename, width, height, res) {
+    if (use_ragg()) {
+        ragg::agg_png(
+            filename = filename,
+            width = width,
+            height = height,
+            res = res
+        )
+    } else {
+        grDevices::png(
+            filename = filename,
+            width = width,
+            height = height,
+            res = res,
+            type = default_device_type()
+        )
+    }
+}
+
+device_svg <- function(filename, width, height) {
+    grDevices::svg(
```

Comment:
We should probably do the same approach for svglite as for ragg? This should result in smaller and higher quality SVG generation.